File: 0015-三数之和/Solution.py
class Solution:
    def threeSum(self, nums):
        if len(nums) < 3:
            return []
        i, j, k = 0, 1, 2
        res = []
        for i in range(len(nums)-2):
            for j in range(i+1, len(nums)-1):
                for k in range(j+1, len(nums)):
                    if nums[i] + nums[j] + nums[k] == 0:
                        res.append([nums[i], nums[j], nums[k]])
        norepeat = []
        for i in res:
            isrepeat = False
            for j in norepeat:
                first, second = sorted(i), sorted(j)
                if first[0] == second[0] and first[1] == second[1]:
                    isrepeat = True
                    break
            if isrepeat:
                continue
            norepeat.append(i)
        
        return norepeat


if __name__ == "__main__":
    print(Solution().threeSum([-4,-2,-2,-2,0,1,2,2,2,3,3,4,4,6,6]))

File: 0015-三数之和/Solution.java
import java.util.ArrayList;
import java.util.Arrays;

class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        if(nums == null || nums.length < 3){ return res; }
        Arrays.sort(nums);
        for (int i = 0; i < nums.length-2; i++) {
            if(i>0 && nums[i]==nums[i-1]){ continue; }

            int left=i+1, right=nums.length-1;
            while(left < right){
                if(left>i+1 && nums[left]==nums[left-1]){
                    left++;
                    continue;
                }
                if(right<nums.length-1 && nums[right]==nums[right+1]){
                    right--;
                    continue;
                }
                int sum = nums[i] + nums[left] + nums[right];
                if(sum == 0){
                    List<Integer> temp = new ArrayList<>();
                    temp.add(nums[i]); temp.add(nums[left]); temp.add(nums[right]);
                    res.add(temp);
                    left++; right--;
                }else if(sum < 0){
                    left++;
                }else{
                    right--;
                }
            }   
        }
        return res;
    }
}